calico/felix/test/test_fetcd_api.py

```diff
@@ -0,0 +1,177 @@
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(1)
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+        self.etcd_api.write_to_etcd.assert_has_calls([status_call,
+                                                      uptime_call])
+
+    def test_update_felix_status_continuous(self):
+        """
+        Test felix status is being continuously updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=5,
+                          REPORTING_TTL_SECS=17)
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(5)
+
+        last_status_call = None
+        for update in range(5):
+            gevent.sleep(5)
+            # Check that both uptime and status were updated
+            self.etcd_api.write_to_etcd.assert_has_calls([status_call, uptime_call])
+            # Update last_status_call and reset write_to_etcd calls
+            last_status_call = self.etcd_api.write_to_etcd.mock_calls[0]
+            self.etcd_api.write_to_etcd.reset_mock()
+
+    @patch('calico.felix.fetcd._reconnect')
+    def test_update_felix_status_reconnects_on_etcd_exception(self, _reconnect):
+        """
+        Test felix status handles exceptions
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=4,
+                          REPORTING_TTL_SECS=12)
+        self.etcd_api.write_to_etcd = Mock(side_effect=EtcdException)
+
+        gevent.sleep(1)
+
+        self.assertTrue(_reconnect.called)
+
+
+class TestIfStatus(object):
+    """
+    Used to check whether status has JSON format and contains expected
+    attributes, whis are:
+    - timestamp in ISO 8601 Zulu format
+    """
+    def __eq__(self, other):
+        import re
+        timestamp_regex = re.compile('.*"status_time": "\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z?".*')
+        has_timestamp =  timestamp_regex.match(str(other)) is not None
+        return is_json(other) and has_timestamp
+    def __repr__(self):
+        return '%s()' % self.__class__.__name__
+
+class TestIfUptime(object):
+    """
+    Used to check whether uptime has correct format (i.e. whether it is
+    non-negative integer)
+    """
+    def __eq__(self, other):
+        is_int = type(other) == int
+        is_non_negative = other >= 0
+        return is_int and is_non_negative
+    def __repr__(self):

```

Comment:
Style nitpick: should be a single newline between methods; same as the method above. I would also suggest adding some parens to make the booleans easier to read; i.e.,

```python
is_int = (type(other) == int)
is_non_negative = (other >= 0)
```